Add taiko hit overlay animations

Joppe27 opened the pull request:
Tried to not make things complicated by advancing the frame manually each beat instead of playing the animation and somehow syncing it up for every object. I hope this is not a problem!

One thing to note: `OnNewBeat`'s `Divisor` is set to 2 in situations where the animation is playing at normal speed instead of the usual 1. This is to make sure `OnNewBeat` is called enough so that a hitobject never enters the playfield in between beats, which somtimes makes animations start too late and looks tacky. This is accounted for in the `animationFrame` calculation which makes it look overcomplicated, but it works :)

https://user-images.githubusercontent.com/56885706/200204706-47a6581f-3629-4e0c-ba60-955a51c2f99f.mp4

smoogipoo committed:
Add some debugging for multiplayer test failures

peppy committed:
Remove commented unbind

Joppe27 committed:
Add taiko hitobject animations

Joppe27 committed:
Update DrawableHit test scene to include hit animations

peppy commented on osu.Game.Rulesets.Taiko/Skinning/Legacy/LegacyCirclePiece.cs:
```diff
@@ -45,19 +55,23 @@ private void load(ISkinSource skin, DrawableHitObject drawableHitObject)
+            // when LegacyCirclePiece is used in places without a score processor (e.g. editor), fallback to 0 combo.
+            if (scoreProcessor != null)
+                currentCombo.BindTo(scoreProcessor.Combo);
+            else
+                currentCombo.Value = 0;
```
Shouldn't be required.

Joppe27 replied:
Good point, removed.

Joppe27 committed:
Remove redundant code

peppy committed:
Fix sprites not displaying in storyboard if filename extension is missing in script

peppy committed:
Hide settings/notifications regardless of `IsPresent` state of new overlay

peppy committed:
Add test coverage of overlay interplay

peppy committed:
Adjust judgement text to be more visible

peppy committed:
Remove triangles skin specific implementation from base `DrawableHit`

peppy committed:
Flash piece when hit

peppy committed:
Merge branch 'master' into editor-background

peppy committed:
Limit editor background dim to 75% maximum

peppy committed:
Merge pull request #20907 from cdwcgt/editor-background

Add ability to toggle background dim in editor

peppy committed:
Add the ability to toggle off hit marker displays in the editor

peppy committed:
Move gameplay configuration to interface to allow editor overriding

peppy committed:
Override combo colour brightness normalisation setting only in editor

Piggey committed:
change `positionText` to display "#?" if tracked score is being 51st on the leaderboard

Piggey committed:
add visual tests

Piggey committed:
Merge branch 'ppy:master' into gameplay-leaderboard-fix

peppy committed:
Remove unused `IHasMainCirclePiece` interface

peppy committed:
Move some non-default skin files to correct namespace

peppy committed:
Apply nullability to osu!taiko skinning classes

peppy committed:
Apply nullability to osu! skinning classes

peppy committed:
Apply nullability to osu!mania skinning classes

peppy committed:
Apply nullability to osu!catch skinning classes

peppy committed:
Apply nullability to skinning support classes

peppy committed:
Apply nullability to `SkinnableDrawable`

peppy committed:
Rename `ISkinComponent` to `ISkinLookup`

peppy committed:
Rename `SkinnableTarget` to `GlobalSkinLookup`

peppy committed:
Add "Component" prefix to lookup naming

peppy committed:
Remove `LookupName` from base `ISkinComponentLookup`

peppy committed:
Fix some lingering inspections

peppy committed:
Fix editor playing too many sounds when user performs a manual seek during playback

swekka committed:
fix inaccurate tablet area dimensions when applying aspect ratio

frenzibyte committed:
Rename test steps

Co-authored-by: Joseph Madamba <madamba.joehu@outlook.com>

Piggey committed:
change displayed text from `#?` to `-`

Piggey committed:
Merge remote-tracking branch 'origin/gameplay-leaderboard-fix' into gameplay-leaderboard-fix

Piggey committed:
fix comment lol

swekka committed:
move tablet area calculations to functions

Joehuu committed:
Merge branch 'master' into close-side-overlays-when-main-overlay

Joehuu committed:
Merge pull request #21166 from peppy/close-side-overlays-when-main-overlay

Closes https://github.com/ppy/osu/issues/21162

smoogipoo committed:
Merge pull request #21183 from peppy/skinnable-nullability

Apply nullability considerations to all skin related classes

smoogipoo committed:
Merge pull request #21185 from peppy/fix-editor-sample-spam

Fix editor playing too many sounds when user performs a manual seek during playback

peppy committed:
Merge branch 'master' into skinnable-things-cleanup

smoogipoo committed:
Merge pull request #21173 from peppy/toggle-hit-markers

Add the ability to toggle off hit marker displays in the editor

smoogipoo committed:
Merge pull request #21168 from peppy/fix-storyboard-extension-lookup

Fix sprites not displaying in storyboard if filename extension is missing in script

smoogipoo committed:
Merge pull request #21167 from peppy/fix-storyboard-variables

Fix some issues with storyboard variables

smoogipoo committed:
Merge pull request #21171 from peppy/unpollute-taiko-drawable-hit

Remove triangles skin specific implementation from base osu!taiko `DrawableHit`

smoogipoo committed:
Merge pull request #21184 from peppy/skinnable-things-cleanup

Skinnable class naming pass

peppy committed:
Merge pull request #21192 from Samaoo/fix-tablet-area-dimensions

Fix inaccurate tablet area dimensions when applying aspect ratio

peppy committed:
Merge branch 'master' into fix-default-24h-time-format

peppy committed:
Update framework

Susko3 committed:
Fix NRT causing CI failure

smoogipoo committed:
Merge pull request #21144 from peppy/fix-storyboard-end-start-time

Fix incorrect handling of storyboard events with `end_time` before `start_time`

smoogipoo committed:
Merge pull request #21147 from peppy/fix-timeline-editor-crash

Fix editor hard crash when beatmap file specified out-of-range timeline zoom value

smoogipoo committed:
Merge pull request #21154 from peppy/seeding-fixes

Fix multiple issues with seeding screen in tournament client

peppy committed:
Fix nullability issue

peppy committed:
Merge pull request #21197 from Susko3/fix-ci-failing

Fix NRT causing CI failure

peppy committed:
Merge pull request #21164 from Susko3/fix-default-24h-time-format

Fix `Prefer24HourTime` default value being wrong

maromalo committed:
Add computeRawScore()

Susko3 committed:
Fix parsing of `Language` when using default system locale

bdach committed:
Merge branch 'master' into improve-ui-control-sound-sync

peppy committed:
Merge pull request #21199 from Susko3/fix-language-parsing

Fix parsing of `Language` when using system default language

smoogipoo committed:
Fix test failure

bdach committed:
Merge branch 'master' into improve-ui-control-sound-sync

smoogipoo committed:
Merge branch 'master' into combo-colour-brightness-limit

bdach committed:
Merge pull request #21121 from peppy/improve-ui-control-sound-sync

Update settings UI animation to better suit immediacy of sound effects

smoogipoo committed:
Merge branch 'master' into taiko-tick-trackiung

smoogipoo committed:
Merge branch 'master' into taiko-barline-separation

smoogipoo committed:
Merge branch 'master' into taiko-judgement-fixes

smoogipoo committed:
Merge pull request #21150 from peppy/taiko-tick-trackiung

Move first tick tracking logic inside `TickPiece`

smoogipoo committed:
Merge pull request #21148 from peppy/taiko-barline-separation

Move major barline portion to default implementation to allow for further customisation

smoogipoo committed:
Merge pull request #21151 from peppy/taiko-judgement-fixes

Fix various issues surrounding hit judgements

peppy committed:
Change default beatmap listing key binding to `Ctrl`+`B`

peppy committed:
Merge pull request #21203 from peppy/change-beamtap-listing-key-vinding

Change default beatmap listing key binding to `Ctrl`+`B`

peppy committed:
Merge branch 'master' into combo-colour-brightness-limit

peppy committed:
Add back removed configuration elements

peppy committed:
Fix one more reverted change

peppy committed:
Merge pull request #20736 from smoogipoo/combo-colour-brightness-limit

Normalise combo colour brightness

peppy committed:
Merge branch 'master' into websocket-chat-2

swekka committed:
fix tablet aspect ratio values

bdach committed:
Merge branch 'master' into argon-taiko

swekka committed:
use MathF

bdach committed:
Adjust taiko argon transformer to new naming

bdach committed:
Fix code inspection

bdach committed:
Rename `{catch -> taiko}Component`

bdach committed:
Fix failing gameplay bindings test

peppy committed:
Merge pull request #21153 from peppy/argon-taiko

First pass osu!taiko "argon" skin

peppy committed:
Merge pull request #21207 from bdach/fix-gameplay-binding-test

Fix failing gameplay bindings test

swekka committed:
address reviews

swekka committed:
remove unused using directive

Tom94 committed:
Target rounded fragment shader

Compatibility with osu-framework change https://github.com/ppy/osu-framework/pull/5512

bdach committed:
Add helper scripts for using local resources

bdach commented on osu.Game.Rulesets.Taiko/Skinning/Legacy/LegacyCirclePiece.cs:
```diff
@@ -96,5 +126,15 @@ private void updateAccentColour()
+            if (!enableAnimations)
+                return;
+
+            animationFrame = beatIndex % 4 == 0 || (beatIndex - 1) % 4 == 0 ? 1 : 0;
```
Do you mind explaining further what this is supposed to be doing? I can't see anything of this sort in stable source.

Joppe27 replied:
This part determines which frame of the animation to show on every `OnNewBeat` call. Because it is called twice as much as usual (`Divisor` is doubled, see PR description for reason), `animationFrame`'s result needs to be the same for two calls in a row. Below image might make this more clear, each x is an `OnNewBeat` call:

![Screenshot 2022-11-12 160358](https://user-images.githubusercontent.com/56885706/201480339-71612088-d5f6-42ab-b845-0d643c14815a.jpg)

If I didn't need to double `Divisor`, I would be able to do this which might be more in line with stable:
```
animationFrame = beatIndex % 2 == 0 ? 1 : 0;
```

bdach replied:
That simplified variant looks closer to what I'd expect to see.

I could have a closer look at this but I'd like these workarounds to not exist if possible, or to be properly signposted if they need to exist. Are you sure that stable doesn't have the same "animation tackiness"?

Joppe27 replied:
To help understand what I meant by "tackiness", notice the animations at the very right edge of the playfield start too late without the `Divisor` workaround applied (this does not happen in stable):
https://user-images.githubusercontent.com/56885706/201521627-5002ff88-b236-457a-bee5-ab769a8b0e8b.mp4

I'm not sure how I would go about solving this, for this to work without a workaround I think the hitobjects would need to be created earlier so they have a bigger margin of time to be affected by `OnNewBeat` before they're visible to the player. I'm not sure if this is possible/wanted though. What are your thoughts on this issue?

bdach replied:
To confirm, this animation discrepancy is what you're trying to fix, correct?

![image](https://user-images.githubusercontent.com/20418176/202872800-5c8cbc56-a24f-48bb-a5d9-d34533d7196b.png)

If so, then it may be a decent idea to ditch `BeatSyncedContainer` and kinda copy stable's implementation a little, which is:

```csharp
            bool doCircleAnimation = ComboCounter.HitCombo >= 50;


            if (doCircleAnimation)
            {
                circlesFrame = (ComboCounter.HitCombo >= 150 ? MouthOpen(2) : MouthOpen(1)) ? 0 : 1;


                hitObjectManager.hitObjectsMinimal.ForEach(ho =>
                {
                    SliderTaiko s = ho as SliderTaiko;
                    if (s != null)
                        s.sliderStartCircle.SpriteHitCircle2.CurrentFrame = circlesFrame;
                    else
                    {
                        HitCircleTaiko h = ho as HitCircleTaiko;
                        if (h != null)
                            h.SpriteHitCircle2.CurrentFrame = circlesFrame;
                    }
                });
            }
```

wherein

```csharp
        internal static bool MouthOpen(int multiplier)
        {
            double beatLength = AudioEngine.BeatLength;
            return AudioEngine.SyncTime % ((beatLength * 2) / multiplier)
                   >= beatLength / multiplier;
        }
```

Note that all of the above is using very basic quantities, i.e. current time and beat length.

Maybe there's a way to move this 1:1 from stable using bits and pieces alreadyh available in lazer. Tough to say.

Joppe27 replied:
Yes, that's what I meant. I was able to port the stable implementation over to lazer without much effort (thanks for giving me the code) and it seems to work nicely. I still needed `LegacyCirclePiece` to be a `BeatSyncedContainer` so that I can easily get `BeatLength`, but apart from that `OnNewBeat` and the `Divisor` workaround have been removed! Let me know if below looks okay since it basically changes this entire PR, ready to push.

``` csharp
protected override void Update()
        {
            base.Update();

            if (currentCombo.Value >= 150)
            {
                enableAnimations = true;
                multiplier = 2;
            }
            else if (currentCombo.Value >= 50)
            {
                enableAnimations = true;
                multiplier = 1;
            }
            else
            {
                enableAnimations = false;
                (foregroundLayer as IFramedAnimation)?.GotoFrame(0);
                return;
            }

            if (BeatSyncSource.ControlPoints != null)
                beatLength = BeatSyncSource.ControlPoints.TimingPointAt(LifetimeStart).BeatLength;

            animationFrame = Time.Current % ((beatLength * 2) / multiplier) >= beatLength / multiplier ? 0 : 1;

            if (enableAnimations)
                (foregroundLayer as IFramedAnimation)?.GotoFrame(animationFrame);
        }
```

bdach replied:
I believe you don't really need `BeatSyncedContainer` for `BeatLength` access, you could just as well resolve `IBeatSyncProvider` directly and thus cut the middleman.

I'd say that doing so and going with the alternate implementation above may be warranted in this particular case, since this is an unorthodox use case, and similar things [have been done in other places already](https://github.com/ppy/osu/blob/master/osu.Game/Storyboards/Drawables/DrawableStoryboardAnimation.cs#L131).

peppy committed:
Merge pull request #21209 from bdach/use-local-resources-scripts

Add helper scripts for using local resources

peppy committed:
Merge pull request #20991 from smoogipoo/websocket-chat-2

Receive chat messages via notification websocket

peppy committed:
Merge branch 'master' into chat-silences

Tom94 committed:
Fix compilation after framework-side PR updates

peppy committed:
Update ack code after incorrect merge

maromalo committed:
Merge branch 'master' into long-standard

peppy committed:
Remove excess logging on `WebSocketNotificaitonsClient`

peppy committed:
Expand on why call needs to be made every 10 minutes

peppy committed:
Simplify ack re-perform flow

peppy committed:
Fix last silence ID being updated too often, causing most silences to be missed

peppy committed:
Add xmldoc for chat ack request

Joppe27 committed:
Replace switch with if-else statements and remove bindable MinValue

Joppe27 committed:
Replace switch with if-else statements and remove bindable MinValue

Joppe27 commented on osu.Game.Rulesets.Taiko/Skinning/Legacy/LegacyCirclePiece.cs:
```diff
@@ -74,6 +85,25 @@ protected override void Update()
+
+            switch (currentCombo.Value)
+            {
+                case >= 150:
```
Fine by me, applied!

Joppe27 commented on osu.Game.Rulesets.Taiko/Skinning/Legacy/LegacyCirclePiece.cs:
```diff
@@ -2,22 +2,32 @@
         private Drawable backgroundLayer = null!;
+        private Drawable? foregroundLayer;
+
+        private Bindable<int> currentCombo { get; } = new BindableInt { MinValue = 0 };
```
Thanks, removed.

Joppe27 committed:
Remove ScoreProcessor dependency and resolve GameplayState instead

Joppe27 committed:
Remove ScoreProcessor dependency and resolve GameplayState instead

Joppe27 commented on osu.Game.Rulesets.Taiko/Skinning/Legacy/LegacyCirclePiece.cs:
```diff
@@ -28,7 +38,7 @@ public LegacyCirclePiece()
 
         [BackgroundDependencyLoader]
-        private void load(ISkinSource skin, DrawableHitObject drawableHitObject)
+        private void load(ISkinSource skin, DrawableHitObject drawableHitObject, ScoreProcessor? scoreProcessor)
```
Seems better indeed, have changed.

peppy committed:
Merge pull request #21073 from smoogipoo/chat-silences

Remove chat messages from silenced users

peppy committed:
Fix `SkinnableSprite` lookups broken in lazer-first skins

Regressed with removal of local `GetTexture` calls in https://github.com/ppy/osu/commit/e19ba65f9186afab21aec536a636d7d152636dde

smoogipoo committed:
Merge pull request #21220 from peppy/fix-skinnable-sprite-lookups

Fix `SkinnableSprite` lookups broken in lazer-first skins

peppy committed:
Update framework

peppy committed:
Update resources

peppy committed:
Merge branch 'master' into only-rounded-shader

peppy committed:
Merge pull request #21208 from Tom94/only-rounded-shader

Target rounded fragment shader

peppy committed:
Add pooling support to `BarHitErrorMeter`

peppy committed:
Apply nullability to `BarHitErrorMeter`

Joehuu committed:
Fix beatmap card expanded content not blocking clicks from behind

peppy committed:
Apply pooling support to `ColourHitErrorMeter`

Susko3 committed:
Fix `FallbackSampleStore.GetAsync` fallback logic

smoogipoo committed:
Merge pull request #21226 from Joehuu/beatmap-card-expanded-block-clicks

Fix beatmap card expanded content not blocking clicks from behind

smoogipoo committed:
Merge pull request #21227 from Susko3/fix-FallbackSampleStore-async-fallback-logic

Fix `FallbackSampleStore.GetAsync` fallback logic

mk56-spn committed:
Exclude sliderticks from the " freeze "

peppy committed:
Ensure flowing animation starts correctly from zero

peppy committed:
Fix number of judgements shown potentially exceeding upper limit

peppy committed:
Remove upper limit for pooling

bdach committed:
Merge pull request #21235 from mk56-spn/Freeze_frame_slidertick_fix

Fix slider ticks appearing before sliders are done snaking in with Freeze Frame

bdach committed:
Merge branch 'master' into hit-error-meter-pooling

peppy committed:
Merge pull request #21240 from peppy/hit-error-meter-pooling

Apply pooling support to hit error metres

EVAST9919 committed:
Add test scene for Triangles

EVAST9919 committed:
Reset triangles on scale change

EVAST9919 committed:
Fix incorrect number of added triangles

EVAST9919 committed:
Improve triangles distribution

peppy committed:
Fix argon swells incorrectly flashing on every hit

bdach committed:
Merge pull request #21243 from peppy/fix-swell-flashing

Fix argon and triangles swells incorrectly flashing on every hit

nanashi-1 committed:
disable certain options when no beatmap is selected

nanashi-1 committed:
add disable

nanashi-1 committed:
AddButton now returns the button

nanashi-1 committed:
visual tests

peppy committed:
Move equilateral triangle ratio to constant

peppy committed:
Add test scene covering taiko swell

nanashi-1 committed:
fixed naming

nanashi-1 committed:
remove redundant code

nanashi-1 committed:
change `disabledAlpha` from `readonly` to `const`

nanashi-1 committed:
add space

nanashi-1 committed:
fix naming

peppy committed:
Fix "reset to full area" button not always working correctly

peppy committed:
Add basic osu!taiko "argon" swell visual

peppy committed:
Fix combo colour normalisation setting not applying to editor test play

peppy committed:
Merge pull request #21242 from EVAST9919/triangles-fix

Triangles background improvements

frenzibyte committed:
Move `IGameplaySettings` override to compose screen to keep test mode unaffected

frenzibyte committed:
Merge branch 'master' into fix-full-area-reset

frenzibyte committed:
Revert beatmap option button state changes

frenzibyte committed:
Disable beatmap options button when none selected

frenzibyte committed:
Revert buttons order change

frenzibyte committed:
Remove leftover constant

smoogipoo committed:
Merge pull request #21257 from peppy/fix-test-mode

Fix combo colour normalisation setting not applying to editor test play

smoogipoo committed:
Merge pull request #21255 from peppy/fix-full-area-reset

Fix "reset to full area" button not always working correctly

peppy committed:
Don't seek to current editor location when location is close to (or before) the first object

peppy committed:
When returning to the editor after test play, use the original editor time rather than the point of exit

smoogipoo committed:
Merge pull request #21256 from peppy/argon-taiko-swell

Add basic osu!taiko "argon" swell visual

peppy committed:
Catch and gracefully handle file/directory enumeration failures during stable import

Closes https://github.com/ppy/osu/issues/21214.

peppy committed:
Use more generic exception type (issue reported `AggegateException`)

smoogipoo committed:
Remove unused using

smoogipoo committed:
Merge pull request #21265 from peppy/editor-test-play-seek-sanity

Don't seek to current editor location when location is close to (or before) the first object

peppy committed:
Fix sliders not correctly taking on full available length after changing curve type

peppy committed:
Move "keybindings" keyword to correct section

Without this, things like tablet settings would show when searching for
bindings, even though these settings have nothing to do with key
bindings.

peppy committed:
Auto-advance binding for ruleset key bindings

nanashi-1 committed:
add condition

Piggey committed:
override `sort()` function in `SoloGameplayLeaderboard`

Piggey committed:
change scoresType to be a bindable

Piggey committed:
add tests

peppy committed:
Eagerly dispose of triangles intro textures to avoid holding for full length of game session

peppy committed:
Mark beatmap listing placeholder textures as large

peppy committed:
Avoid loading beatmap listing placeholders until required

peppy committed:
Avoid loading overlay headers until first open

Also switches them to use non-atlased target as they won't benefit much
from atlasing.

peppy committed:
Fix news post images never unloading from memory after first display

Piggey committed:
revert changes from `GameplayLeaderboard` tests

Piggey committed:
oh oops

Piggey committed:
cleanup

Piggey committed:
Merge branch 'master' into gameplay-leaderboard-fix

peppy committed:
Fix supporter filters text not being set correctly

peppy committed:
Fix `NowPlayingOverlay` loading background texture too early (and permanently)

smoogipoo committed:
Merge pull request #21277 from peppy/beatmap-listing-texture-reduction

Reduce unnecessary texture overhead incurred by beatmap listing

smoogipoo committed:
Merge pull request #21278 from peppy/overlay-header-texture-reduction

Avoid loading overlay headers until first open

smoogipoo committed:
Merge pull request #21279 from peppy/news-load-unload

Fix news post images never unloading from memory after first display

smoogipoo committed:
Merge pull request #21276 from peppy/eager-triangles-disposal

Eagerly dispose of triangles intro textures to avoid holding for full length of game session

smoogipoo committed:
Merge pull request #21284 from peppy/now-playing-overlay-texture-reduction

Fix `NowPlayingOverlay` loading background texture too early (and permanently)

peppy committed:
Merge pull request #21244 from nanashi-1/beatmapoptions-disabled

Disable certain beatmap option buttons when there are no beatmaps selected

frenzibyte committed:
Fix beatmap options test failure due to no beatmap being selected

peppy committed:
Fix potential incorrect connection state resulting in null reference

smoogipoo committed:
Merge pull request #21286 from frenzibyte/fix-test-failure

Fix beatmap options test failure due to no beatmap selected

smoogipoo committed:
Refactor a bit for readability

smoogipoo committed:
Merge pull request #21269 from peppy/fix-key-bindings-keyword

Move "keybindings" keyword to correct section

smoogipoo committed:
Merge pull request #21268 from peppy/fix-path-change-expected-length

Fix sliders not correctly taking on full available length after changing curve type

peppy committed:
Reduce time waited on a score submission token from 60 to 30s

smoogipoo committed:
Compute raw score in-line

Saves on an allocation (though it seems to only be a ValueType so it
shouldn't be a big deal) and a tail-call.

peppy committed:
Fix API queue only being flushed once while in a failing state

smoogipoo committed:
Merge branch 'master' into long-standard

peppy committed:
Fix triangle button flash effect looking incorrect

smoogipoo committed:
Merge pull request #21287 from peppy/fix-null-ref-persistent-endpoint-connector

Fix potential incorrect connection state resulting in null reference

peppy committed:
Fix being able to bind two non-modifier keys to the same binding

smoogipoo committed:
Merge pull request #21288 from peppy/fix-no-network-slow-gameplay-load

Fix entering gameplay while network connection is in a bad state taking too long

smoogipoo committed:
Merge pull request #21140 from maromalo/long-standard

Standardise scores to `long` type

smoogipoo committed:
Merge pull request #21289 from peppy/fix-triangle-button-flash

Fix triangle button flash effect looking incorrect

peppy committed:
Add more test coverage to `TestSceneKeyBindingPanel`

peppy committed:
Add failing test coverage to `TestSceneKeyBindingPanel` for multiple non-modifiers being bound

peppy committed:
Merge pull request #21290 from smoogipoo/multiplayer-test-debug

Add some debugging for multiplayer test failures

frenzibyte committed:
Fix timeline potentially scrolling at extents while not dragging

peppy committed:
Merge branch 'master' into gameplay-leaderboard-fix

frenzibyte committed:
Suppress nullable warning for now

peppy committed:
Move constant to a better location

peppy committed:
Fix inspections / typos

peppy committed:
Tidy up bindable flow

peppy committed:
Redirect through validity function rather than overriding `Sort`

peppy committed:
Update test to split out per-type tests

peppy committed:
Fix initial visual state of positions before `sort` runs

peppy committed:
Make new test more resilient to false passes

smoogipoo committed:
Merge pull request #21293 from frenzibyte/fix-timeline-extent-scrolling

Fix timeline potentially scrolling at extents while not dragging

smoogipoo committed:
Merge pull request #21292 from peppy/fix-binding-two-keys-single-binding

Fix being able to bind two non-modifier keys to the same binding

smoogipoo committed:
Merge pull request #21270 from peppy/auto-advance-bindings

Auto-advance binding for ruleset key bindings

Piggey committed:
Merge branch 'ppy:master' into gameplay-leaderboard-fix

peppy committed:
Merge pull request #21191 from Piggey/gameplay-leaderboard-fix

Display - symbol when tracked in-game leaderboard score is below 50

peppy committed:
Fix excess string interpolation

frenzibyte committed:
Fix "perform from screen" not considering screen load state

frenzibyte committed:
Add test coverage

bdach committed:
Update tests to reflect desired behaviour

peppy committed:
Merge pull request #21299 from frenzibyte/perform-from-runner-load-state

Fix "perform from screen" invoking action on non-loaded screens

frenzibyte committed:
Fix editor not always playing hitsounds with clock offsets applied

frenzibyte committed:
Hide scroll speed slider on rulesets which don't support it

frenzibyte committed:
Fix intermittent present beatmap test failures

peppy committed:
Merge pull request #21306 from frenzibyte/fix-editor-seeking-state

Fix editor not always playing hitsounds with clock offsets applied

peppy committed:
Merge pull request #21310 from frenzibyte/fix-present-beatmap-test

Fix intermittent present beatmap test failures

peppy committed:
Merge pull request #21267 from peppy/catch-file-enumeration-errors

Catch and gracefully handle file/directory enumeration failures during stable import

frenzibyte committed:
Remove unused using

frenzibyte committed:
Merge branch 'master' into disable-scroll-speed-on-osu

bdach committed:
Merge branch 'master' into editor-test-play-dont-retain-test-time

bdach committed:
Merge pull request #21266 from peppy/editor-test-play-dont-retain-test-time

When returning to the editor after test play, use the original editor time rather than the point of exit

bdach committed:
Merge branch 'master' into fix-tablet-aspect-ratio-values

bdach committed:
Merge pull request #21205 from Samaoo/fix-tablet-aspect-ratio-values

Fix incorrect rounding of tablet aspect ratio values

frenzibyte committed:
Introduce `IDrawableScrollingRuleset` for editor consumption

frenzibyte committed:
Identify scrolling rulesets using temporary `DrawableRuleset` instances

frenzibyte committed:
Fix timing screen test scene not creating editor beatmap properly

bdach committed:
Remove unnecessary `#nullable disable`

bdach committed:
Merge pull request #21307 from frenzibyte/disable-scroll-speed-on-osu

Hide scroll speed control from effects section on rulesets which don't support it

Joppe27 committed:
Remove all BeatSyncedContainer elements and calculate frame in Update()

Joppe27 committed:
Update tests to support changes and clean up my old/bad code

Joppe27 committed:
Merge branch 'taiko-hitobject-animations' of https://github.com/Joppe27/osu into taiko-hitobject-animations

Joppe27 committed:
Love me some merge conflicts

Joppe27 commented:
OH NO, I did something really really wrong with git

peppy commented:
@Joppe27 yes it's fine to force push after rebasing on the latest `master`.

Joppe27 commented:
I rebased on upstream master, but github desktop still shows "commit not pushed to remote repository" (the up arrow) on commits that are not mine, the same thing that led to this mess. Is it okay to force push like this or will the same issue happen? I don't know why it's showing these commits as local.. Maybe to avoid wasting your time any longer with my git help desk questions it's better to just close this PR and make a new one from a new working branch?
![image](https://user-images.githubusercontent.com/56885706/203122333-62a72387-2c72-4bd9-ae45-6e93c652ade9.png)

bdach commented:
I'm not familiar with github desktop UI but that does not look right no. How are you rebasing even?

There are other ways to salvage this mess but closing and opening afresh sounds fine/easiest.

Joppe27 commented:
Closing because of the problems with this branch.
